Update to scrapper.py

PxINKY opened the pull request:
Added edge case if the creator has left the platform / no longer exists
Added edge case if download_url doesn't exist
Added edge case if product was gifted


Examples of this:
A deleted Gumroad account will leave the product/download available but `result["product"]["creator"]["name"]` would return `None`.

Gumroad added file embeds directly into the page, allowing creators to embed videos/images into the page, this result removes `download_url` from the item object.

If the user received the item using Gumroads gifting system they can't generate payment info.

PxINKY committed:
Update scrapper.py

Patched creator in the case that the creator has left the platform / no longer exists

PxINKY committed:
Update scrapper.py

Catch for text file embeds

PxINKY committed:
Update scrapper.py

fixed download URL being none

PxINKY committed:
Update scrapper.py

Patched out invalid characters in directory names for window machines

PxINKY committed:
Update scrapper.py

Fixed an issue with gifted items

obsessedcake commented on gumroad_utils/scrapper.py:
```diff
@@ -227,7 +236,8 @@ def _traverse_tree(items: list[dict], tree_path: Path, parent_folder: Path) -> N
             for item in items:
-                if item["type"] != "file":
+                # NOTE(PxINKY) Gumroad added a "file" type that's embedded into the page resulting in no download_url
+                if item["type"] != "file" or item["download_url"] is None:
```
Could you please add log, which file doesn't have a download URL,  here?

PxINKY replied:
Here are some examples i found:

**Google drive / External file host**
`{'type': 'file', 'file_name': 'Air Captain.spp - Google Drive', 'description': '', 'extension': 'URL', 'file_size': None, 'pagelength': None, 'duration': None, 'id': 'DhuqduTkneMdU1ksrgK5Kg==', 'download_url': None, 'stream_url': None, 'kindle_data': None, 'latest_media_location': None, 'content_length': None, 'read_url': None, 'external_link_url': 'https://drive.google.com/file/d/(Hiding extension due to Terms of Service)', 'subtitle_files': [], 'pdf_stamp_enabled': False, 'processing': False, 'thumbnail_url': None}`

**URL Embed**
`{'type': 'file', 'file_name': 'Avatars-3.0-Manager', 'description': '', 'extension': 'URL', 'file_size': None, 'pagelength': None, 'duration': None, 'id': 'SpLhtW9tbOdjB0eeq71Giw==', 'download_url': None, 'stream_url': None, 'kindle_data': None, 'latest_media_location': None, 'content_length': None, 'read_url': None, 'external_link_url': 'https://github.com/VRLabs/Avatars-3.0-Manager', 'subtitle_files': [], 'pdf_stamp_enabled': False, 'processing': False, 'thumbnail_url': None}`

**Youtube Embed**
`{'type': 'file', 'file_name': 'How to set up your hairstyle v2 ⭐ (Easy setup) - By AidenFur - YouTube', 'description': '', 'extension': 'URL', 'file_size': None, 'pagelength': None, 'duration': None, 'id': 'cM1_iVZaiUMamkMKZkaz_A==', 'download_url': None, 'stream_url': None, 'kindle_data': None, 'latest_media_location': None, 'content_length': None, 'read_url': None, 'external_link_url': 'https://www.youtube.com/watch?v=kAypVPhdUis&lc=Ugy8UnmwuaOeVQtU-Nt4AaABAg', 'subtitle_files': [], 'pdf_stamp_enabled': False, 'processing': False, 'thumbnail_url': None}`

That being said you defiantly could check/use `external_link_url`

obsessedcake commented:
Hi! Thanks a lot for your patch! Glad to see new improvements to this script :+1:

obsessedcake commented:
> Gumroad added file embeds directly into the page, allowing creators to embed videos/images into the page, this result removes download_url from the item object.

In this case we can try to extract download URL from raw HTML, can't we?

---

I've made a  small utility method from [f30e7ef/gumroad_utils/scrapper.py](https://github.com/obsessedcake/gumroad-utils/blob/f30e7efe02f13fca13350664a250eb029409c202/gumroad_utils/scrapper.py) to extract alll files from raw HTML.

```py
    def _get_all_files(self, soup: BeautifulSoup) -> dict[str, str]:    
        raw_files: dict[str, str] = {}
        for file in soup.find_all("div", attrs={"role": "treeitem"}, class_="js-file-list-element"):
            file_type = file.select_one("li:nth-child(1)").string.lower()
            file_name = file.select_one("h4").string
            file_url = self._session.base_url + file.select_one("a", href=True)["href"]

            raw_files[f"{file_name}.{file_type}"] = file_url
```

`{file_name}.{file_type}` - not the best key here because we can have many files with same name in different folders. If I remember correctly, each file should have a unique id in JSON. If they use this "uid" in HTML then it's an easy go, otherwise meh...

---

Anyway this patch is already good, so this can be fixed in a separate PR!

PxINKY committed:
Update scrapper.py

Added sanitize_filename method
This is for window machines to prevent invalid characters from being used when building directory/file names

PxINKY committed:
Update scrapper.py

Fixed hardcoded replacement (`self._slash_replacement`) I accidentally left it.

PxINKY committed:
Update scrapper.py

Fixed an oversight on my end
